src/cbmc/cbmc_parse_options.cpp

```diff
@@ -879,6 +880,8 @@ bool cbmc_parse_optionst::process_goto_program(
   link_to_library(
     goto_model, log.get_message_handler(), cprover_c_library_factory);
 
+  add_malloc_may_fail_variable_initializations(goto_model);
```

Comment:
If this is now part of the correct set up for goto programs can you either A. add this to all such set-up code ( goto-analyzer, goto-diff, goto-instrument at least) or B. add it to one of the existing methods -- I would not oppose it being part of `link_to_library` as it is part of the handling of the library.

Reply:
It definitely can’t go into `link_to_library` and shouldn’t go into any of the instrumentation tools. The reason this is put in this location is specifically to _avoid_ instrumentation code setting these variables, they should _only_ be set by cbmc. Arguably jbmc and goto-analyzer too, but _definitely_ not goto-instrument etc.

Reply:
To summarise above, I don’t really agree from what I understand of the situation, but maybe I’m missing something.

Reply:
They shouldn't just be `cbmc` models; `goto-analyzer` will need them as well.  They may not work that well at the moment but let's not make them any more broken than they currently are.  In `jbmc` I don't know what their models are written in, Java or C.  It would be good if it was possible for it to work.

My concern is avoiding divergence in how the tools set up and configure goto programs.  This has been a source of subtle bugs in the past.

Reply:
@martin-cs Again, I _could_ put this code in goto-analyzer but I still don’t quite understand what good it’ll do. The models library as a matter of fact is currently not being used for goto-analyzer (and as far as I know this has been the case for _at least_ 5 years), so if I add this code it’ll naturally go untested. So the problem is that then I’m adding code to goto-analyzer that is vestigial to begin with (i.e. literally a no-op in all circumstances). While I do understand that we want to keep goto transforms consistent between different tools in _this_ particular case I don’t see the value in it. Right now goto-analyzer  doesn’t even understand things like `__CPROVER_allocate` etc (or rather, I don’t think any of the domains do). For me the problem here is I’m adding code to places that I _know_ does nothing right now, and that I can’t be sure will do the right thing whenever someone changes the surrounding context in such a way that would allow it to do something. Whenever someone in the future would read that goto-analyzer code and asks "what is this bit of the code for", the true answer would be "it does absolutely nothing because the preconditions for it doing anything are not there". If it ever starts doing anything we’ll have no way of knowing it’s doing the right thing because we didn’t (and couldn’t) test it until then.

I think we’re both concerned about maintainability here, but we seem to disagree on which way round is better for maintainability.

@peterschrammel @piotr-grabalski @NlightNFotis @thomasspriggs would you mind chipping in? Very much willing to be overruled here, but I want to be sure we’re doing the right thing here.

Reply:
@martin-cs Would you be OK with it if I made a different PR adding these changes to goto-analyzer and we can have the discussion on whether or not it _should_ go there on that PR? Getting this fix in for cbmc at least is somewhat time critical for us.

Reply:
Yes, I think it should be called by goto-analyzer too.
IMO, this shouldn't be implemented through the models library at all, but by a pass over the goto model - but that may follow afterwards. This PR is for urgently fixing a major bug in CBMC when used with goto-cc. I.e. priority is getting the bug fixed today. Tomorrow we can look how to improve the solution.

Reply:
The main reason I’d want it to be in the models library is because IMHO whether or not this can return NULL is part of the model, and you should be able to figure out what behaviour the models library has from looking at the library code. Instrumenting logic means distributing the logic between the models library source files and different parts of the C++ code base, but like you said we should discuss how to move forward with this when we’ve actually fixed the problem.

Reply:
@hannes-steffenhagen-diffblue would you be OK adding a comments (and commented out code?) in the other locations saying why it shouldn't / isn't / may be needed?  Would be happy to approve that and get it merged in an hour or so when the CI passes?

@peterschrammel I feel the library solution has a number of advantages; should we talk about this at some point?

Reply:
@thomasspriggs agreed -- we do need to discuss this.  Adding the models and configuring them are two different things and yes, we do need to think about it.  Do you want to open an issue to discuss this?